### PR TITLE
Check QoS policy when configuring intraprocess, skip interprocess publish when possible

### PR DESCRIPTION
**Pull request by ivanpauno**
Follow up of https://github.com/ros2/rclcpp/pull/664/commits
Which was unintentionally merged into https://github.com/ros2/rclcpp/pull/667/

### CODE REVIEW

**Comment by wjwwood** [comment]
I think by "over" you mean "into", "over" sounds like "as opposed to" or "instead of"?

File: rclcpp/include/rclcpp/node.hpp [hunk 1]
```diff
@@ -64,6 +64,17 @@
 namespace rclcpp
 {
 
+/// Used as argument in create_publisher and create_subscriber.
+enum class IntraprocessSetting
+{
+  /// Explicitly enable intraprocess comm at publisher/subscription level.
+  ENABLE,
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node.hpp]: Let's not use ALL_CAPS, as that is often used by macros, and could collide here. (imagine someone did `#define ENABLE ...`, which is not unthinkable.

Instead, let's do `Enable`, `Disable`, and `Default` (or `DefaultFromNode` but that doesn't look very nice to me, another option is maybe `NodeDefault`).

File: rclcpp/include/rclcpp/node_impl.hpp [hunk 2]
```diff
@@ -81,17 +83,30 @@ template<typename MessageT, typename Alloc, typename PublisherT>
...
 
+  bool use_intra_process;
+  switch (use_intra_process_comm) {
+    case IntraprocessSetting::ENABLE:
+      use_intra_process = true;
+      break;
+    case IntraprocessSetting::DISABLE:
+      use_intra_process = false;
+      break;
+    default:  // IntraprocessSetting::DEFAULT_FROM_NODE
```
Comment (wjwwood) [inline rclcpp/include/rclcpp/node_impl.hpp]: This should be its own case instead, and an error should be thrown if yet something else is used.

File: rclcpp/src/rclcpp/publisher.cpp [hunk 3]
```diff
@@ -231,6 +230,19 @@ PublisherBase::setup_intra_process(
   IntraProcessManagerSharedPtr ipm,
   const rcl_publisher_options_t & intra_process_options)
 {
+  // Skip intraprocess configuration if the "durability" qos policy is not "volatile".
+  // TODO(ivanpauno):
+  //    RMW_QOS_POLICY_DURABILITY_SYSTEM_DEFAULT is for FastRTPS and Connext also
+  //    equal to "volatile". It would be good to add a
+  //    is_durability_policy_volatile(rcl_publisher_options_t options)
+  //    method to rcl, and corresponging methods in rmw and the rmw vendor
+  //    implementations, which will also checks if "system_default" is "volatile" or not.
```
Comment (wjwwood) [inline rclcpp/src/rclcpp/publisher.cpp]: But isn't this exactly what your other pr's are aiming to do?

Comment (ivanpauno) [reply rclcpp/src/rclcpp/publisher.cpp]: I will update when I merge them

File: rclcpp/include/rclcpp/node.hpp (hunk 1 above)
Comment (ivanpauno) [reply rclcpp/include/rclcpp/node.hpp]: Up to now, most of the "enum class" in rclcpp follow the ALL_CAPS style, except of `CallbackGroupType`. I agree we should only use it in macros, so I will change it.

Comment (wjwwood) [reply rclcpp/include/rclcpp/node.hpp]: Yeah, there might be other cases that should change, but let's arc towards what we think is right.

File: rclcpp/src/rclcpp/publisher.cpp (hunk 3 above)
Comment (ivanpauno) [reply rclcpp/src/rclcpp/publisher.cpp]: Sorry that I wasn't clear. But I'm not thinking to merge this before the other PRs. I'm now going to update this.

**Comment by ivanpauno** [comment]
CI:
* Linux [Build Status](http://ci.ros2.org/job/ci_linux/6585/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/2923/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/5402/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/6405/)

**Comment by ivanpauno** [comment]
[Here] is the characterization of rclcpp comm with this improvement. It's working as expected.
I've also tried the intraprocess demos, and I've checked that interprocess subscriptions receive messages as expected if the publisher is configured with `use_intra_process(true)` option.
@wjwwood Do you think this is ready to be merged?

**Comment by wjwwood** [comment]
I left some comments on the word document, if you could address them. But that document doesn't really apply to this pull request (it is needed, but doesn't tell us if this pr is ready or not).

**Comment by ivanpauno** [comment]
> I left some comments on the word document, if you could address them. But that document doesn't really apply to this pull request (it is needed, but doesn't tell us if this pr is ready or not).

Yes sure. I ran it to check if this improves performance or not.
I'm merging now. I will end addressing the document comments tomorrow.

**Pull request merged**